cpp/src/reductions/reductions.cpp

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/reduction.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include "reduction_functions.hpp"
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+// Allocate storage for a single identity scalar
+std::unique_ptr<scalar> make_identity_scalar(
```

Comment:
I'd make this call a type dispatched function that returns a default constructed `scalar_type_t<Element>`.

Reply:
For `make_string_scalar`, the first argument is different.

Reply:
I'm not following. The default ctor doesn't take an arguments.

Reply:
I tried both of these
`return std::unique_ptr<scalar>(new cudf::experimental::scalar_type_t<T>{});`
`return std::unique_ptr<scalar>(new cudf::experimental::scalar_type_t<T>());`

for both of them, is_valid = true. `is_valid` should be `false`.

Reply:
Why would you expect a default constructed `string_view` to be null?

Reply:
not only `string_view`, for other types too, scalar is not null. I expected it to be zero initialized https://en.cppreference.com/w/cpp/language/zero_initialization

Reply:
Why would you expect _any_ type to be null when it is default constructed?

Reply:
> I tried both of these
> `return std::unique_ptr<scalar>(new cudf::experimental::scalar_type_t<T>{});`
> `return std::unique_ptr<scalar>(new cudf::experimental::scalar_type_t<T>());`
> 
> for both of them, is_valid = true. `is_valid` should be `false`.

`new T()` or `new T{}` is zero initialized right? 
default constructed is `new T` which I didn't call here.

Reply:
> new T() or new T{} is zero initialized right?

No, this is default constructed. 

This is zero-initialization:
```
T obj{}; // value-init (which calls default ctor)

T obj = T{}; // zero init
```

Reply:
created `make_default_constructed_scalar(type)` factory function.